core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java

```diff
@@ -142,4 +115,75 @@ private void doExecute(BlockingRunnable runnable) {
          blockedTasks.offer(runnable);
       }
    }
+
+   private class ControllerThread extends Thread {
+
+      private final LinkedList<BlockingRunnable> readyList;
+      private final Semaphore semaphore;
+      private volatile boolean interrupted;
+
+      public ControllerThread(String controllerThreadName) {
+         super(controllerThreadName);
+         readyList = new LinkedList<>();
```

Comment:
I normally prefer ArrayDeque over LinkedList since we aren't adding/removing elements from the middle